Updating `dice` example based on Updated Specification for the reference application

lightsta1ker opened the pull request:
Addresses https://github.com/open-telemetry/opentelemetry-go-contrib/issues/7937

lightsta1ker committed:
Updating uninstrumented  example

lightsta1ker committed:
Updating instrumented example

pellared commented on examples/dice/instrumented/main.go:
```diff
@@ -80,10 +86,55 @@ func newHTTPHandler() http.Handler {
 	return handler
 }
+
+func handleRollDice(w http.ResponseWriter, r *http.Request) {
```
Can we move this to `rolldice.go`?

pellared commented on examples/dice/uninstrumented/main.go:
```diff
@@ -60,8 +67,52 @@ func newHTTPHandler() http.Handler {
 	return mux
 }
+
+func handleRollDice(w http.ResponseWriter, r *http.Request) {
```
Can we move this to `rolldice.go`?

lightsta1ker committed:
chore: lint and suggestions

lightsta1ker committed:
Merge branch 'main' into dice/issue-7937

lightsta1ker committed:
chore: rebase & refactor suggestions

pellared commented on examples/dice/instrumented/rolldice.go:
```diff
@@ -19,42 +21,141 @@ import (
+	// Check if rolls is a number.
+	rolls, err := strconv.Atoi(rollsParam)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
```
what about Content-Type?

lightsta1ker replied:
Thought of leaving it to the defaults as it's a simple app and not explicitly specified in the reference. But I will set `application/json` as that's how the app responds.

lightsta1ker commented on examples/dice/instrumented/rolldice.go:
```diff
@@ -19,42 +21,141 @@ import (
+	if rolls <= 0 {
+		err := errors.New("rolls must be positive")
+		span.RecordError(err)
+		logger.ErrorContext(ctx, "error", "error", err)
```
Ah, Yes. The caller does log the same error at line 100

`logger.ErrorContext(r.Context(), err.Error())`

I think, this line was from earlier that escaped refactor. Will update.

lightsta1ker committed:
chore: refactor suggestions

lightsta1ker committed:
Merge branch 'main' into dice/issue-7937

lightsta1ker committed:
fix: Limiting memory allocation for rolls

flc1125 commented on examples/dice/instrumented/rolldice.go:
```diff
@@ -19,42 +22,145 @@ import (
+	// Parse query parameters.
+	rollsParam := r.URL.Query().Get("rolls")
+	player := r.URL.Query().Get("player")
+	maxRolls := 1000 // Arbitrary limit to prevent Slice memory allocation with excessive size value.
```
From a performance perspective, this is a good configuration.

However, I am considering that since the specification does not mention this situation and it is just an example.

Therefore, extreme cases can be temporarily ignored. I recommend removing it to comply with the specification.

flc1125 commented on examples/dice/instrumented/rolldice.go:
```diff
@@ -19,42 +22,145 @@ import (
+	if err != nil || rolls > maxRolls {
+		// Signals invalid input (<=0).
+		w.WriteHeader(http.StatusInternalServerError)
+		logger.ErrorContext(r.Context(), err.Error())
```
Since the condition `if err != nil || rolls > maxRolls {` implies that `err` might be `nil`.

Of course, once `maxRolls` is removed, this issue is naturally resolved.

flc1125 reviewed (commented):
The normative review is completed, with additional notes:

According to the specification requirements, support for the `OTLP` exporter is needed. Currently, this part is not yet supported, but I believe it can be fully implemented through a separate PR.

lightsta1ker committed:
Merge branch 'main' into dice/issue-7937

lightsta1ker committed:
chore: consolidating rollDice observability logic

lightsta1ker committed:
chore: refactoring suggestion

lightsta1ker commented:
@flc1125 I thought those checks were not relevant to this PR as it's not customer facing.

lightsta1ker committed:
Merge branch 'main' into dice/issue-7937

lightsta1ker committed:
fix: lint fixes & error handling

pellared commented on examples/dice/instrumented/rolldice.go:
```diff
@@ -19,42 +22,149 @@ import (
+		return
+	}
+
+	if rolls > maxRolls {
```
Can we move this validation to `rollDice`?

lightsta1ker replied:
sure

lightsta1ker commented on examples/dice/uninstrumented/rolldice.go:
```diff
@@ -4,26 +4,94 @@
+	}
+}
+
+// rolldice is the outer function which does the error handling.
```
Again, its a reference to the doc - "The library should have an outer function that does the error handling as described above." from initial iteration. Will update.

lightsta1ker committed:
fix: refactoring suggestions

lightsta1ker committed:
chore: format with gofumpt

lightsta1ker committed:
Merge branch 'main' into dice/issue-7937

pellared commented on examples/dice/instrumented/rolldice.go:
```diff
@@ -19,42 +21,141 @@ import (
-		msg = "Anonymous player is rolling the dice"
+		logger.DebugContext(r.Context(), "player=%s rolled %v", player, results)
+	}
+	logger.InfoContext(r.Context(), " %s %s -> 200 OK", r.Method, r.URL.String())
```
@svrnm, I will try to followup separately from this PR so that we keep going forward.

lightsta1ker committed:
Merge branch 'main' into dice/issue-7937

lightsta1ker committed:
chore: rebasing & error handling changes

lightsta1ker committed:
Merge branch 'main' into dice/issue-7937

lightsta1ker committed:
Merge branch 'main' into dice/issue-7937

pellared committed:
Merge branch 'main' into dice/issue-7937

pellared commented:
@open-telemetry/go-approvers, anyone has some time to review this PR?

dmathieu commented on examples/dice/instrumented/rolldice.go:
```diff
@@ -19,42 +22,147 @@ import (
+
+	lastRollsGauge, err = meter.Int64ObservableGauge(
+		"dice.last.rolls",
+		metric.WithDescription("The last rolls value observed"),
```
I think neither are great 😬 


```suggestion
		metric.WithDescription("The most recent rolled value"),
```

lightsta1ker committed:
Update examples/dice/instrumented/rolldice.go

Co-authored-by: Damien Mathieu <42@dmathieu.com>

pellared committed:
Merge branch 'main' into dice/issue-7937

svrnm reviewed (approved):
from what I understand this follows the specification I provided, so I can approve it from that point of view :-)

svrnm commented:
Question for @open-telemetry/go-approvers:

https://github.com/lightsta1ker/opentelemetry-go-contrib/blob/dice/issue-7937/examples/dice/instrumented/go.mod

```
replace (
	go.opentelemetry.io/contrib/bridges/otelslog => ../../../bridges/otelslog
	go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp => ../../../instrumentation/net/http/otelhttp
)
```

Can this be removed and by that the example be made independent of the source repository?

dmathieu commented:
The examples aren't using unreleased APIs. So the replace statements aren't strictly required.

svrnm commented:
> The examples aren't using unreleased APIs. So the replace statements aren't strictly required.

then, let's drop it, so people can run this example without the need to take anything else from this repo (or more importantly build anything from this repo).

lightsta1ker committed:
Update examples/dice/instrumented/otel.go

Co-authored-by: Robert Pająk <pellared@hotmail.com>

lightsta1ker committed:
chore: using otel-codes

lightsta1ker committed:
chore: resolving conflicts

lightsta1ker committed:
fix: merge conflict

lightsta1ker committed:
chore: golangci-lint fix

svrnm commented:
@lightsta1ker are the conflicts new ones? can you resolve them so we can make progress here?

lightsta1ker committed:
chore: resolving conflicts

lightsta1ker committed:
chore: lint issue fix

lightsta1ker commented:
> @lightsta1ker are the conflicts new ones? can you resolve them so we can make progress here?

Hi @svrnm , resolved conflicts & fixed lint issues.

svrnm commented:
@open-telemetry/go-maintainers anything missing here to get this merged?:-)

lightsta1ker committed:
Merge branch 'main' into dice/issue-7937

svrnm commented:
@open-telemetry/go-maintainers ping:)

dashpole commented:
I see a bunch of unresolved comment threads.  Most look actionable to me.

dashpole commented on examples/dice/uninstrumented/rolldice.go:
```diff
@@ -4,26 +4,100 @@
+	results, err := rolldice(rolls)
+	if err != nil {
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusInternalServerError)
```
This probably isn't required for a demo app

dashpole commented on examples/dice/instrumented/rolldice.go:
```diff
@@ -19,42 +22,163 @@ import (
+	_, span := tracer.Start(ctx, "rollOnce")
+	defer span.End()
+
+	roll := 1 + rand.IntN(6) //nolint:gosec // G404: Use of weak random number generator (math/rand instead of crypto/rand) is ignored as this is not security-sensitive.
```
nit-pick, but this would be good to do

dashpole commented on examples/dice/instrumented/rolldice.go:
```diff
@@ -19,42 +22,163 @@ import (
+	}
+
+	results := make([]int, rolls)
+	for i := range rolls {
```
nit, but it would be good to have consistency

lightsta1ker committed:
chore: fixing lint issues & addressed comments

lightsta1ker commented:
> I see a bunch of unresolved comment threads. Most look actionable to me.

Hi @dashpole,
I've addressed the comments & fixed lint failure as well.
Thank you

dashpole commented on examples/dice/uninstrumented/rolldice.go:
```diff
@@ -4,26 +4,101 @@
+	}
+
+	results := make([]int, rolls)
+	for i := range rolls {
```
Incorrect.  You can range over an int.

dashpole commented on examples/dice/uninstrumented/rolldice.go:
```diff
@@ -4,26 +4,101 @@
+		return
+	}
+
+	results, err := rolldice(rolls)
+	if err != nil {
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusInternalServerError)
+		log.Printf("ERROR: %v", err)
+		return
```
Not that important.

dashpole commented on examples/dice/instrumented/rolldice.go:
```diff
@@ -19,42 +23,172 @@ import (
+
+	lastRollsGauge, err = meter.Int64ObservableGauge(
+		"dice.last.rolls",
+		metric.WithDescription("The most recent rolled value"),
```
I agree this is inconsistent.  Is this following some sort of predefined schema for the examples?

lightsta1ker replied:
No, It's not following any schema. I agree with the description update as well.
